### PR TITLE
Add unfused SBHD compressed sparse attention

### PR DESCRIPTION
**Pull request by FDecaYed**
# Add unfused SBHD compressed sparse attention

## Summary

Add the reference SBHD Compressed Sparse Attention implementation needed by DeepSeek-V4, while keeping ordinary DSA on the current-main implementation and API.

Related to #5795.

## Scope

- Add SBHD compressor, indexer, compressed-position handling, sparse/dense attention, sliding-window attention, and attention-sink behavior in `csa.py`.
- Add CSA configuration fields and focused forward/backward/parity coverage.
- Reuse selective-FP32 projection support from #5929 and safe MLA RoPE primitives from #5944.
- Extend the shared DSA teacher helper with a detached non-compressed log-sum-exp input so CSA loss uses the full compressed + window + sink denominator required by #5960, without changing ordinary DSA behavior.
- Reject explicit attention masks rather than silently crossing padding or document boundaries.
- Use explicit `ProcessGroupCollection` plumbing.

## Supported scope

BF16, eager, SBHD, fixed-length, TP1/PP1/CP1.

## Non-goals

- Fused/cuDNN CSA kernels; that is #6404.
- Packed THD, context parallelism, pipeline parallelism, or CUDA-graph integration.
- Top-level DeepSeek-V4 orchestration and logging; that is #6402.
- HybridModel layer construction; that is #6405.
- GPTModel support or recipes.

## Review boundary and dependencies

#5929 and #5944 are merged in `main`. This branch contains one residual commit: `10ae87b793306371657c41ffc2ee4880c842907a`.

Please review the complete PR diff: **6 files, +2,460/-9**.

Current-main ordinary DSA remains the source of truth. #6020 defines the main-first DSA versus dev-only CSA ownership boundary, while #5960 supplies the corrected full-denominator teacher semantics incorporated here. #5795 remains historical feature and test provenance, not the patch base.

## Validation

Passed locally on the final changeset:

- clean worktree;
- valid GPG signature and required `Signed-off-by` trailer;
- `git diff --check`;
- Python parsing and Ruff 0.9.10 on changed Python files;
- isort 5.13.2 on newly import-touched files;
- Black 26.3 on attributable new hunks.

The macOS host has no usable Torch/CUDA runtime, so no local GPU/runtime result is claimed. Full functional CI passed on exact head `10ae87b793306371657c41ffc2ee4880c842907a`: [run 32114744957](https://github.com/NVIDIA/Megatron-LM/actions/runs/32114744957) (`L1`, 5 repeats, non-lightweight; 151 jobs succeeded, 8 expected skips, 0 failures, and the final `Nemo_CICD_Test` gate passed). No dependency or lock-file change is part of this commit.

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]

**Comment by guihong-nv** [comment]
@FDecaYed I think we can move the files under `megatron/core/transformer/experimental_attention_variant` to `megatron/core/ops/attention/dsa,mla (depending on if the file is for dsa or mla)` or `megatron/core/transformer/experimental_attention_variant` for now (we split them later). Please also add some tags that clarify if a kernel is deterministic or not; it's okay to put an `unknown` here.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 1]
```diff
@@ -0,0 +1,914 @@
...
+    # --- Gather KV at topk positions ---
+    # kv_full: [n_kv, b, hn] -> [b, n_kv, hn]
+    kv_t = kv_full.permute(1, 0, 2)
+
+    safe_indices = topk_indices.clamp(min=0).long()  # [b, sq, topk]
+    safe_indices_exp = safe_indices.unsqueeze(-1).expand(-1, -1, -1, hn)  # [b, sq, topk, hn]
+    # [b, n_kv, hn] -> [b, 1, n_kv, hn] -> gather -> [b, sq, topk, hn]
+    kv_gathered = torch.gather(
+        kv_t.unsqueeze(1).expand(-1, sq, -1, -1), dim=2, index=safe_indices_exp
+    )
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: Thanks — fixed in `d02991d`. The sparse-attention gather now flattens batch and KV position to `[b * n_kv, hn]`, applies per-batch offsets, and uses `index_select`, so backward accumulates into the compact KV buffer instead of materializing `[b, sq, n_kv, hn]`. I added a CUDA regression test covering `b=2`, repeated indices, and `-1` padding, with forward and query/KV/sink gradients checked against a loop oracle; it also explicitly rejects `torch.gather`. The exact-head full functional CI passed.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 2]
```diff
@@ -0,0 +1,914 @@
...
+        score = score + self.ape.view(1, ratio, 1, -1)
+
+        if self.overlap:
+            kv = self._overlap_transform(kv, fill_value=0)
+            score = self._overlap_transform(score, fill_value=float("-inf"))
+
+        weights = torch.softmax(score, dim=1, dtype=torch.float32).to(kv.dtype)
+        kv = (kv * weights).sum(dim=1)  # [n_compressed, b, head_dim]
+
+        kv = self.norm(kv.to(x.dtype))
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: Thanks — fixed in `d02991d`. Compressor pooling now keeps the softmax weights, `kv.float() * weights`, and the reduction in FP32, then casts the pooled result once to `x.dtype` before normalization. The new BF16-input test checks the forward result and both KV/score gradients exactly against an FP32 pooling oracle. The exact-head full functional CI passed.

File: megatron/core/transformer/transformer_config.py [hunk 3]
```diff
@@ -350,6 +350,22 @@ class TransformerConfig(ModelParallelConfig):
...
+    csa_window_size: int = 128
+    """Sliding window size for compressed sparse attention."""
+
+    csa_compress_ratios: Optional[List[int]] = None
+    """Per-layer compress ratios, e.g. [0, 0, 4, 128, 4, 128, ...]."""
+
+    csa_compress_rotary_base: float = 40000.0
+    """RoPE base for compressed KV positions in compressed sparse attention."""
+
+    csa_dense_mode: bool = False
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/transformer_config.py:368]: Addressed in `d02991d`. I added an explicit TODO above these two fields. After the follow-on slices were recut, #6402—not #6405—is the actual consumer: it selects the per-layer compression ratio and constructs the compressed-KV rotary embedding. The comment also records that neither field has a production reader in this primitive-only PR.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 4]
```diff
@@ -0,0 +1,914 @@
...
+    """Sparse core attention for CompressedSparseAttention.
+
+    Combines sliding window attention with compressed KV attention.  The spec always
+    provides compressor and indexer submodule specs; this ``__init__`` inspects
+    ``config.csa_compress_ratios[layer_idx]`` and conditionally builds them:
+
+    * ``ratio == 0``:  window-only (compressor and indexer NOT built)
+    * ``ratio == 4``:  window + 4x compressed + learned Indexer (both built)
+    * ``ratio == 128``: window + 128x compressed, attend to all (compressor built only)
+    """
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py:730]: Addressed in `d02991d`. I rewrote the docstring to match the actual constructor contract: `compress_ratio` is caller-provided; ratios `<= 1` are window-only; ratios `> 1` build the compressor; and ratio 4 builds the indexer only when dense mode is disabled. It no longer claims to read `config.csa_compress_ratios` or presents 0/4/128 as an exhaustive dispatch table.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 5]
```diff
@@ -0,0 +1,914 @@
...
+        self.n_local_heads = config.num_attention_heads
+
+        if softmax_scale is None:
+            softmax_scale = config.v_head_dim**-0.5
+        self.softmax_scale = softmax_scale
+
+        # Learnable attention sink per head, kept in reference-checkpoint FP32.
+        self.attn_sink = mark_keep_in_fp32(
+            nn.Parameter(torch.zeros(self.n_local_heads, dtype=torch.float32))
+        )
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py]: Addressed in `d02991d`. Since this native SBHD slice explicitly supports TP1 only, I chose the scoped rejection path rather than introducing partially sharded sink semantics. The constructor now rejects `tp_size != 1` with a clear error, `n_local_heads` was renamed to `num_attention_heads`, and both the module boundary and direct attention helper validate the query/sink head count. Focused tests cover TP rejection, module-level head mismatch, and direct sink mismatch.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -514,6 +519,52 @@ def compute_dsa_indexer_loss(
...
+
+def _normalize_indexer_teacher_target(
+    target: torch.Tensor, non_compressed_lse: torch.Tensor | None
+) -> torch.Tensor:
+    """L1-normalize teacher mass without changing the legacy DSA path."""
+    if non_compressed_lse is None:
+        return dsa_indexer_loss.normalize_indexer_target(target)
+    return target / target.sum(dim=-1, keepdim=True).clamp_min(torch.finfo(torch.float32).tiny)
+
+
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:595]: Thanks — fixed in `d02991d`. The external-mass path now separates the conditional distribution over compressed keys from each head’s compressed mass, keeps the latter in log space, and applies one common per-row shift across heads when the absolute mass would underflow; that shift cancels in the final L1 normalization. This path deliberately does not use `INDEXER_LOSS_EPS`: any positive compressed mass is normalized, while an exactly zero row stays zero. The manual KL gradient is now `predict * target.sum(-1) - target`, so zero-mass rows produce zero gradient. Added regressions for a 2-head/2-key `LSE≈1000` float64 oracle, fully masked finite-zero target/autograd, valid zero-mass manual gradients, and unchanged moderate/legacy behavior. The exact-head full functional CI passed.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 7]
```diff
@@ -0,0 +1,914 @@
...
+                if self.training and torch.is_grad_enabled():
+                    q_indexer, k_indexer, weights_indexer = self.indexer.forward_before_topk(
+                        x_det, qr_det
+                    )
+                    indexer_loss_coeff = self.config.dsa_indexer_loss_coeff or 0.0
+                    key_for_loss = compressed_kv.unsqueeze(2).expand(-1, -1, np, -1)
+                    weights_for_unfused = weights_indexer.float() * self.indexer.softmax_scale
+                    non_compressed_lse = _compute_unfused_csa_non_compressed_lse(
+                        query, kv, self.attn_sink, window_idxs, self.softmax_scale
+                    )
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py:918]: Thanks — this optimization is intentionally deferred in the native reference path. I added an explicit TODO documenting the duplicate window-logit/gather work, that #6404’s fused training backend avoids it, and that the native fallback should later share the gathered window KV/logits. This PR keeps the correctness-first implementation unchanged.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 8]
```diff
@@ -0,0 +1,914 @@
...
+                            num_layers=self.config.num_layers + (self.config.mtp_num_layers or 0),
+                        )
+                else:
+                    _, topk_indices_compressed = self.indexer(x_det, qr_det, mask=causal_mask)
+
+                n_valid_per_pos = positions // self.compress_ratio  # [sq, 1]
+                valid = (topk_indices_compressed >= 0) & (topk_indices_compressed < n_valid_per_pos)
+                compress_topk_idxs = torch.where(
+                    valid, topk_indices_compressed + offset, torch.tensor(-1, device=x.device)
+                )
```
[the remark this answers is not in the document]
Comment (FDecaYed) [reply megatron/core/transformer/experimental_attention_variant/csa.py:957]: Fixed in `d02991d`. The compressed causal mask and per-position valid counts now use bounded `@lru_cache` helpers keyed by shape/device inputs; the cached mask is batch-expanded as a view, and `torch.where` uses the Python scalar `-1`. Unit coverage verifies mask/count semantics and cache reuse.

**Review by guihong-nv — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
